ngraph/core/src/op/gather.cpp

```diff
@@ -128,6 +128,185 @@ shared_ptr<Node> op::v1::Gather::clone_with_new_inputs(const OutputVector& new_a
+        NODE_VALIDATION_CHECK(this,
+                              batch_dims <= axis,
+                              "The batch_dims <= axis. But instead got: batch_dims = ",
+                              batch_dims,
+                              ", axis = ",
+                              axis);
+
+        if (data_rank.is_static())
+        {
+            NODE_VALIDATION_CHECK(this,
+                                  axis >= 0 && axis < data_rank.get_length(),
+                                  "The axis must be => 0 and < data_rank. But instead got axis = ",
+                                  axis,
+                                  " data_rank = ",
+                                  data_rank.get_length());
+        }
+    }
+
+    if (indices_rank.is_static() && batch_dims >= 0)
+    {
+        NODE_VALIDATION_CHECK(
+            this,
+            batch_dims < indices_rank.get_length(),
+            "The batch_dims must be < indices_rank. But instead got: batch_dims = ",
+            batch_dims,
+            ", indices_rank = ",
+            indices_rank.get_length());
+    }
+
+    if (data_rank.is_static() && indices_rank.is_static() && batch_dims >= 0)
+    {
+        auto out_rank = data_rank.get_length() + indices_rank.get_length() - 1 - batch_dims;
+        std::vector<Dimension> result_dims(out_rank);
+        PartialShape output_pshape(result_dims);
+
+        // implementation of out_shape formula
+        // data.shape[:batch_dims] + data.shape[batch_dims:axis] + indices.shape[batch_dims:] +
+        // data.shape[axis + 1:]
+        int i = 0;
+        for (; i < batch_dims; i++)
+        {
+            NODE_VALIDATION_CHECK(this,
+                                  data_pshape[i].compatible(indices_pshape[i]),
+                                  "Shapes ",
+                                  data_pshape,
+                                  " and ",
+                                  indices_pshape,
+                                  " are not consistent. data and indices must have equal or "
+                                  "intersecting sizes until batch_dims");
+
+            output_pshape[i] = data_pshape[i] & indices_pshape[i];
+        }
+
+        if (is_axis_set())
+        {
+            int64_t axis = get_axis();
+            for (; i < axis; i++)
+            {
+                output_pshape[i] = data_pshape[i];
+            }
+            for (; i < axis + indices_rank.get_length() - batch_dims; i++)
+            {
+                output_pshape[i] = indices_pshape[batch_dims - axis + i];
+            }
+            for (; i < out_rank; i++)
+            {
+                output_pshape[i] = data_pshape[batch_dims + 1 - indices_rank.get_length() + i];
+            }
+        }
+
+        set_output_type(0, data_type, output_pshape);
+    }
+    else if (data_rank.is_static() && indices_rank.is_static() && batch_dims < 0)
+    {
+        // batch_dims < 0 could be only if axis is not set and in in optimistic scenario
```

Comment:
"in in optimistic scenario"? We always assume that the graph is in correct state and all parameters/inputs of the nodes have valid values.